### PR TITLE
Refactor tests

### PR DESCRIPTION
**Pull request by huangqinjin**
@mstorsjo Please feel free to push more commits on this PR.

### CODE REVIEW

File: test/test-cmake.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    -DCMAKE_BUILD_TYPE=RelWithDebInfo
+    -DCMAKE_SYSTEM_NAME=Windows
+)
+
+case $OSTYPE in
+    darwin*)
+        CMAKE_ARGS+=(
+            -DCMAKE_EXE_LINKER_FLAGS=/MANIFEST:NO
+            -DCMAKE_SHARED_LINKER_FLAGS=/MANIFEST:NO
+            -DCMAKE_MSVC_DEBUG_INFORMATION_FORMAT=Embedded
```
Comment (mstorsjo) [inline test/test-cmake.sh:30]: Why are these arguments needed on Darwin? I presume that means that msvctricks.exe doesn't work there (breaking the mt.exe return code rewrite, and leaving cmake/ninja waiting on the mspdbsrv background process)? A comment would probably be good.

**Comment by mstorsjo** [comment]
I think this looks mostly reasonable. The recursive use of `test.sh` is a bit mind-twisting at first, but I think I see the reasons for it.

File: test/test-cmake.sh (hunk 1 above)
Comment (huangqinjin) [reply test/test-cmake.sh:30]: https://github.com/mstorsjo/msvc-wine/pull/65#issuecomment-1531747302

Comment (mstorsjo) [reply test/test-cmake.sh:30]: Oh, ok, thanks for reminding me! A comment mentioning these reasons would be good.

Comment (huangqinjin) [reply test/test-cmake.sh:33]: Done.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.13.4)
+cmake_minimum_required(VERSION 3.25)
 project(hello)
 
 set(hello_SOURCES hello.c)
```
Comment (huangqinjin) [inline test/CMakeLists.txt:4]: @mstorsjo Is the `WIN32` condition necessary? Is this cmakelist supposed to target platforms other than Windows?

Comment (mstorsjo) [reply test/CMakeLists.txt:4]: I'd like the test to remain compileable for any platform, yeah. It's not something that we'd do much, admittedly, but I'd like the cmakelist to remain like a real one would be for a proper cross platform project.

**Comment by huangqinjin** [comment]
It seems there is a problem with Cmake when targeting ARM, though it doesn't fail.

https://github.com/mstorsjo/msvc-wine/actions/runs/5269443375/jobs/9527585853#step:5:129

**Comment by mstorsjo** [comment]
> It seems there is a problem with Cmake when targeting ARM, though it doesn't fail.
> 
> https://github.com/mstorsjo/msvc-wine/actions/runs/5269443375/jobs/9527585853#step:5:129

It seems like CMake doesn't pass `/EHsc` when compiling C++ code targeting ARM, contrary to other architectures.

**Comment by mstorsjo** [comment]
> I think this PR is ready and if it looks good to you, please **squash and merge**.

It looks reasonable to me, thanks! I pushed a minor cosmetic update to the branch. Although to me, the commits look neatly separated, so to me I'd actually prefer to just do a rebase+merge here, if you don't mind.

**Comment by huangqinjin** [comment]
> prefer to just do a rebase+merge here

I'm OK with that.

File: test/test-cmake.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        ) ;;
+esac
+
+EXEC "" CC=${BIN}cl CXX=${BIN}cl RC=${BIN}rc cmake -S"$TESTS" -GNinja "${CMAKE_ARGS[@]}"
+EXEC "" ninja -v
+
+# Rerun ninja to make sure that dependencies aren't broken.
+EXEC ninja-rerun ninja -d explain -v
+DIFF ninja-rerun.out - <<EOF
+ninja: no work to do.
```
Comment (mstorsjo) [inline test/test-cmake.sh:48]: If this testcase fails, the explanation is stored in `ninja-rerun.err`, but that's not printed anywhere. It'd be nice if that was printed, especially if the diff test failed.

A second note; the diff is a bit hard to read when it's reversed (the actual output is the original state and the expected output is the new state, in the diff); it's easy to change by swapping the sides in the diff into `DIFF - ninja-rerun.out`, but that makes it slightly harder to read when the output name doesn't line up with the `EXEC` line above.

**Pull request merged**